cmd/signature-v2.go

```diff
@@ -290,13 +290,7 @@ func canonicalizedResourceV2(encodedPath string, encodedQuery string) string {
 			canonicalQueries = append(canonicalQueries, key)
 			continue
 		}
-		// Resources values should be unescaped
-		unescapedVal, err := url.QueryUnescape(val)
-		if err != nil {
-			errorIf(err, "Unable to unescape query value (query = `%s`, value = `%s`)", key, val)
-			continue
-		}
-		canonicalQueries = append(canonicalQueries, key+"="+unescapedVal)
+		canonicalQueries = append(canonicalQueries, key+"="+val)
```

Comment:
Okay now i understood what you are trying to do, its a regression fix. I was able to reproduce this problem. But this is not a complete fix, removing this brings back another bug. To fix both of them here is diff that you need to send.

```diff
diff --git a/cmd/signature-v2.go b/cmd/signature-v2.go
index f38ec73..f646e31 100644
--- a/cmd/signature-v2.go
+++ b/cmd/signature-v2.go
@@ -92,13 +92,14 @@ func doesPresignV2SignatureMatch(r *http.Request) APIErrorCode {
                encodedQuery = tokens[1]
        }
 
-       queries := strings.Split(encodedQuery, "&")
-       var filteredQueries []string
-       var gotSignature string
-       var expires string
-       var accessKey string
-       var err error
-       for _, query := range queries {
+       var (
+               filteredQueries []string
+               gotSignature    string
+               expires         string
+               accessKey       string
+               err             error
+       )
+       for _, query := range strings.Split(encodedQuery, "&") {
                keyval := strings.Split(query, "=")
                switch keyval[0] {
                case "AWSAccessKeyId":
@@ -108,18 +109,17 @@ func doesPresignV2SignatureMatch(r *http.Request) APIErrorCode {
                case "Expires":
                        expires, err = url.QueryUnescape(keyval[1])
                default:
-                       unescapedQuery, qerr := url.QueryUnescape(query)
-                       if qerr == nil {
+                       var unescapedQuery string
+                       unescapedQuery, err = url.QueryUnescape(query)
+                       if err == nil {
                                filteredQueries = append(filteredQueries, unescapedQuery)
-                       } else {
-                               err = qerr
                        }
                }
-               // Check if the query unescaped properly.
                if err != nil {
-                       errorIf(err, "Unable to unescape query values", queries)
+                       errorIf(err, "Unable to unescape query %s from (%s)", query, encodedQuery)
                        return ErrInvalidQueryParams
                }
+
        }
 
        // Invalid access key.
@@ -217,7 +217,17 @@ func doesSignV2Match(r *http.Request) APIErrorCode {
                encodedQuery = tokens[1]
        }
 
-       expectedAuth := signatureV2(r.Method, encodedResource, encodedQuery, r.Header)
+       var filteredQueries []string
+       for _, query := range strings.Split(encodedQuery, "&") {
+               unescapedQuery, err := url.QueryUnescape(query)
+               if err != nil {
+                       errorIf(err, "Unable to unescape query %s from (%s)", query, encodedQuery)
+                       return ErrInvalidQueryParams
+               }
+               filteredQueries = append(filteredQueries, unescapedQuery)
+       }
+
+       expectedAuth := signatureV2(r.Method, encodedResource, strings.Join(filteredQueries, "&"), r.Header)
        if v2Auth != expectedAuth {
                return ErrSignatureDoesNotMatch
        }
@@ -267,7 +277,7 @@ func canonicalizedAmzHeadersV2(headers http.Header) string {
 }
 
 // Return canonical resource string.
-func canonicalizedResourceV2(encodedPath string, encodedQuery string) string {
+func canonicalizedResourceV2(encodedQuery string) string {
        queries := strings.Split(encodedQuery, "&")
        keyval := make(map[string]string)
        for _, query := range queries {
@@ -280,6 +290,7 @@ func canonicalizedResourceV2(encodedPath string, encodedQuery string) string {
                }
                keyval[key] = val
        }
+
        var canonicalQueries []string
        for _, key := range resourceList {
                val, ok := keyval[key]
@@ -290,19 +301,12 @@ func canonicalizedResourceV2(encodedPath string, encodedQuery string) string {
                        canonicalQueries = append(canonicalQueries, key)
                        continue
                }
-               // Resources values should be unescaped
-               unescapedVal, err := url.QueryUnescape(val)
-               if err != nil {
-                       errorIf(err, "Unable to unescape query value (query = `%s`, value = `%s`)", key, val)
-                       continue
-               }
-               canonicalQueries = append(canonicalQueries, key+"="+unescapedVal)
-       }
-       if len(canonicalQueries) == 0 {
-               return encodedPath
+               canonicalQueries = append(canonicalQueries, key+"="+val)
        }
-       // the queries will be already sorted as resourceList is sorted.
-       return encodedPath + "?" + strings.Join(canonicalQueries, "&")
+
+       // The queries will be already sorted as resourceList is sorted, if canonicalQueries
+       // is empty strings.Join returns empty.
+       return strings.Join(canonicalQueries, "&")
 }
 
 // Return string to sign for authz header calculation.
@@ -326,9 +330,14 @@ func signV2STS(method string, encodedResource string, encodedQuery string, heade
                headers.Get("Content-Type"),
                headers.Get("Date"),
                canonicalHeaders,
-       }, "\n") + canonicalizedResourceV2(encodedResource, encodedQuery)
+       }, "\n")
+
+       canonicalResource := canonicalizedResourceV2(encodedQuery)
+       if canonicalResource != "" {
+               return stringToSign + encodedResource + "?" + canonicalResource
+       }
 
-       return stringToSign
+       return stringToSign + encodedResource
 }
 
 // Return string to sign for pre-sign signature calculation.
@@ -352,6 +361,11 @@ func presignV2STS(method string, encodedResource string, encodedQuery string, he
                headers.Get("Content-Type"),
                expires,
                canonicalHeaders,
-       }, "\n") + canonicalizedResourceV2(encodedResource, encodedQuery)
-       return stringToSign
+       }, "\n")
+
+       canonicalResource := canonicalizedResourceV2(encodedQuery)
+       if canonicalResource != "" {
+               return stringToSign + encodedResource + "?" + canonicalResource
+       }
+       return stringToSign + encodedResource
 }
```